### PR TITLE
Take advantage of Python 3.12

### PR DESCRIPTION
**Pull request by bmerry**
This is the low-hanging fruit of NGC-1380. I haven't yet used task groups or exception groups, because that's going to require deeper thinking about what the code is doing (plus reading up about the feature). I also haven't changed the generics to use PEP 695 because mypy doesn't support it yet.

Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (n/a) If modules are added/removed: use `sphinx-apidoc -efo doc/ src/` to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (n/a) If qualification tests are changed: attach a sample qualification report
- [x] (n/a) If design has changed: ensure documentation is up to date
- [x] (n/a) If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match

### CODE REVIEW

**Comment by bmerry** [comment]
There's a lot here, but the bulk of it is in the first commit which is run by machine (pyupgrade).

File: qualification/report/generate_pdf.py [hunk 1]
```diff
@@ -670,7 +670,7 @@ def collate_results(results: Iterable[Result]) -> list[ResultSet]:
     for result_list in result_lists.values():
         result_set = ResultSet(result_list)
         # Check whether the CBF titles alone give unique subtitles
-        cbf_titles = set(result.cbf.title for result in result_list if result.cbf is not None)
+        cbf_titles = {result.cbf.title for result in result_list if result.cbf is not None}
```
Comment (james-smith-za) [inline qualification/report/generate_pdf.py:673]: Interesting new syntax.

File: requirements-dev.txt [hunk 2]
```diff
@@ -40,7 +40,6 @@ async-timeout==4.0.3
     # via
     #   -c requirements.txt
     #   aiokatcp
```
Comment (james-smith-za) [inline requirements-dev.txt:42]: Looks like it's still in requirements.txt based on this?
(the aiokatcp dependency can probably wait for another day)

File: qualification/report/generate_pdf.py (hunk 1 above)
Comment (bmerry) [reply qualification/report/generate_pdf.py:673]: Actually that's been around a long time, although I can't immediately find which Python version introduced it.

File: requirements-dev.txt (hunk 2 above)
Comment (bmerry) [reply requirements-dev.txt:42]: Yes, it's only removed as a direct requirement. I expect it to still be pulled in indirectly for the foreseeable future.

File: qualification/report/generate_pdf.py (hunk 1 above)
Comment (bmerry) [reply qualification/report/generate_pdf.py:673]: Looks like it dates right back to Python 3.0 (and 2.7).

**Review by james-smith-za — APPROVED** [review]

**Pull request merged**